packages/squiggle-lang/src/rescript/Reducer/Reducer_Expression/Reducer_Expression.res

```diff
@@ -0,0 +1,90 @@
+module BuiltIn = Reducer_Dispatch_BuiltIn
+module ExpressionValue = ReducerInterface.ExpressionValue
+module Extra = Reducer_Extra
+module MathJs = Reducer_MathJs
+module Result = Belt.Result
+module T = Reducer_Expression_T
+open Reducer_ErrorValue
+
+type expression = T.expression
+type expressionValue = ExpressionValue.expressionValue
+
+/*
+  Shows the Lisp Code as text lisp code
+*/
+let rec show = expression =>
+  switch expression {
+  | T.EList(aList) =>
+    `(${Belt.List.map(aList, aValue => show(aValue))
+      ->Extra.List.interperse(" ")
+      ->Belt.List.toArray
+      ->Js.String.concatMany("")})`
+  | EValue(aValue) => ExpressionValue.show(aValue)
+  }
+
+let showResult = codeResult =>
+  switch codeResult {
+  | Ok(a) => `Ok(${show(a)})`
+  | Error(m) => `Error(${Js.String.make(m)})`
+  }
+
+/*
+  Converts a MathJs code to Lisp Code
+*/
+let parse_ = (expr: string, parser, converter): result<expression, errorValue> =>
+  expr->parser->Result.flatMap(node => converter(node))
+
+let parse = (mathJsCode: string): result<expression, errorValue> =>
+  mathJsCode->parse_(MathJs.Parse.parse, MathJs.ToExpression.fromNode)
+
+module MapString = Belt.Map.String
+type bindings = MapString.t<unit>
+let defaultBindings: bindings = MapString.fromArray([])
+// TODO Define bindings for function execution context
+
+/*
+  After reducing each level of code tree, we have a value list to evaluate
+*/
+let reduceValueList = (valueList: list<expressionValue>): result<expressionValue, 'e> =>
+  switch valueList {
+  | list{EvSymbol(fName), ...args} => (fName, args->Belt.List.toArray)->BuiltIn.dispatch
+  | _ => valueList->Belt.List.toArray->ExpressionValue.EvArray->Ok
+  }
+
+/*
+  Recursively evaluate/reduce the code tree
+*/
+let rec reduceExpression = (expression: expression, bindings): result<expressionValue, 'e> =>
+  switch expression {
+  | T.EValue(value) => value->Ok
+  | T.EList(list) => {
+      let racc: result<list<expressionValue>, 'e> = list->Belt.List.reduceReverse(Ok(list{}), (
+        racc,
+        each: expression,
+      ) =>
+        racc->Result.flatMap(acc => {
+          each
+          ->reduceExpression(bindings)
+          ->Result.flatMap(newNode => {
+            acc->Belt.List.add(newNode)->Ok
+          })
+        })
+      )
+      racc->Result.flatMap(acc => acc->reduceValueList)
```

Comment:
I find the above ~12 lines fairly hard to read, but I'm not very used to reducers. Not sure if a cleanup is possible, no worries if it doesn't seem straightforward

Reply:
That code requires training training training... If it were Haskell then it would be a simple do block. As far as I know there is no simpler way in reScript. Result.flatMap corresponds to Haskell's >>(bind) operator. Imagine we need to write a Haskell do block without the do block but with the bind operator.

do
   a<-sth1
   b<-sth2 a
   sth3

becomes
   sth1 >> ( \a -> (sth2 a  >> (\b -> sth3) )  )

in Haskell.

Now. Result is a Monad. Yet we don't have a do-block in reScript. 
- The question is how to make  reScript equivalent of sth1 >> ( \a -> (sth2 a  >> (\b -> sth3) )  ) simpler.
- The standard code formatter makes that kind of code in reScript even more cryptic. I hate it.

Reply:
Makes sense. The lack of a do-block in reScript is definitely annoying.

Your above comment would be good to make into a comment in that part of the code, for future people who see it. 
"Note: This would be cleaned up with the Haskell do block, but we don't have that."

Reply:
I would support a helper function that hides `>>` notation. 

`let mSequence3 = (sth1, sth2, sth3) => ...`

Reply:
RESOLVED